### DISCUSSION TITLE
When trying to execute `select 1` under read only transaction I do not expect SAWarning (beginning from 2.0.32)

### DISCUSSION THREAD

**Discussion by tomaszkingukrol** [sqlalchemy/sqlalchemy#11964] [Usage Questions]
Starting with version 2.0.32 when trying to run:

```
from sqlalchemy import Engine, create_engine, event, text
from sqlalchemy.orm import Session

engine = create_engine("postgresql://", echo=True)


def engine_connect(conn):
    conn.execute(text("SET TRANSACTION READ ONLY"))  # current ...
    conn.execute(text("SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY"))  # ... and next transactions


event.listen(Engine, "engine_connect", engine_connect)

with Session(engine) as session:
    session.execute(text("SELECT 1"))
```
I got:
```
SAWarning: The engine provided as bind produced a connection that is already in a transaction. This is usually caused by a core event, such as 'engine_connect', that has left a transaction open. The effective join transaction mode used by this session is 'rollback_only'. To silence this warning, do not leave transactions open
  session.execute(text("SELECT 1"))
```

Is this something that should be fixed or is this expected behaviour for some reason?

**Comment by CaselIT** [comment]
Hi,

The warning is correct. Your `engine_connect` event executes sql statements but does not commit o rollback them. This leaves a transaction open in the connection.

What the warning is telling you is that the session will be in the rollback_only modality, detailed here: https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode

**Reply by tomaszkingukrol** [reply]
Thanks for answer. One way to achive read only session is:

`engine = create_engine("postgresql://", execution_options={"postgresql_readonly": True})`

But, could you give some advice how to do that after creation engine without this special option?

**Reply by CaselIT** [reply]
Commit the connection before returning from your `engine_connect` event handler.

**Reply by tomaszkingukrol** [reply]
Many thanks, it works.
